File: sort.py
import random
import time
import sys

def merge_sort(nums, low, high):
	if low < high:
		mid = (low + high) // 2
		merge_sort(nums, low, mid)
		merge_sort(nums, mid + 1, high)
		merge(nums, low, mid, high)
		
def merge(nums, low, mid, high):
	left_len = (mid - low) + 1
	right_len = high - mid
	left = []
	right = []
	
	for index in range(low, mid + 1):
		left.append(nums[index])
	for index in range(mid + 1, high + 1):
		right.append(nums[index])
		
	left_index = 0
	right_index = 0
	num_index = low
	while left_index < left_len and right_index < right_len:
		if left[left_index] < right[right_index]:
			nums[num_index] = left[left_index]
			left_index += 1
		else:
			nums[num_index] = right[right_index]
			right_index += 1
		num_index += 1
	while left_index < left_len:
		nums[num_index] = left[left_index]
		left_index += 1
		num_index += 1
	while right_index < right_len:
		nums[num_index] = right[right_index]
		right_index += 1
		num_index += 1

def create_random_number_array(n):
	random_number_array = []
	for index in range(n):
		random_number_array.append(random.randint(-n, n))
	return random_number_array
	
def main():

	size = int(sys.argv[1])
	
	test_array = create_random_number_array(size)

	start = time.time_ns()
	merge_sort(test_array, 0, size - 1)
	stop = time.time_ns()

	print("Python mergeSort() took", (stop-start)/1000000000, "seconds to execute for", size,"elements")
	
if __name__ == "__main__":
	main()

File: sort.c
#include<stdio.h>
#include <time.h>
#include <stdlib.h>


void merge(int[], int, int ,int);


void mergeSort(int nums[], int low, int high){
	if(low<high){
		int mid = (low+high)/2;
		mergeSort(nums, low, mid);
		mergeSort(nums, mid+1, high);
		merge(nums, low, mid, high);
	}
}


void merge(int nums[], int low, int mid, int high){
	//1. create two arrays
	int left_len = (mid-low)+1;
	int right_len = high-mid;
	int left[left_len];
	int right[right_len];
	
	//2. copy elements from original array into new arrays
	for(int index=0; index<left_len; index++){
		left[index] = nums[low+index];
	}
	
	int right_start = mid+1;
	for(int index=0; index<right_len; index++){
		right[index] = nums[right_start+index];
	}
	
	//3. start the merge process
	int left_index= 0;
	int right_index= 0;
	int num_index= low;
	
	while(left_index<left_len && right_index<right_len){
		if( left[left_index] < right[right_index] ){
			nums[num_index] = left[left_index];
			left_index++;
		}
		else{
			nums[num_index] = right[right_index];
			right_index++;
		}
		num_index++;
	}
	
	while(left_index < left_len){
		nums[num_index] = left[left_index];
		num_index++;
		left_index++;
	}
	
	while(right_index<right_len){
		nums[num_index] = right[right_index];
		num_index++;
		right_index++;
	}
	
}


int* makeArray(int length){
	int* pointer = malloc(sizeof(int)*length);
	//int* pointer = (int*)malloc(sizeof(int)*length);
	for(int i=0;i<length;i++){
    	pointer[i]=(rand()%length);
    }
    return pointer;
}


int main(int argc, char* argv[]){
	int size = atoi(argv[1]);
	
	int* nums = makeArray(size);
	
    clock_t begin = clock();

	mergeSort(nums, 0, size-1);
	clock_t end = clock();
	double timeTaken = (double)(end - begin) / CLOCKS_PER_SEC;
   
    printf("c mergeSort() took %f seconds to execute for %d elements\n", timeTaken, size);
	
    return 0;
}
